core/vm/program.go

```diff
@@ -415,98 +386,163 @@ func jitCalculateGasAndSize(env Environment, contract *Contract, instr instructi
 		// 2. From a non-zero value address to a zero-value address (DELETE)
 		// 3. From a non-zero to a non-zero                         (CHANGE)
 		if common.EmptyHash(val) && !common.EmptyHash(common.BigToHash(y)) {
-			g = params.SstoreSetGas
+			gas = SstoreSetGas64
 		} else if !common.EmptyHash(val) && common.EmptyHash(common.BigToHash(y)) {
 			statedb.AddRefund(params.SstoreRefundGas)
 
-			g = params.SstoreClearGas
+			gas = SstoreClearGas64
 		} else {
-			g = params.SstoreResetGas
+			gas = SstoreResetGas64
 		}
-		gas.Set(g)
 	case SUICIDE:
 		if !statedb.IsDeleted(contract.Address()) {
 			statedb.AddRefund(params.SuicideRefundGas)
 		}
 	case MLOAD:
-		newMemSize = calcMemSize(stack.peek(), u256(32))
+		newMemSize, sizeFault = calcMemSize(stack.peek(), u256(32))
 	case MSTORE8:
-		newMemSize = calcMemSize(stack.peek(), u256(1))
+		newMemSize, sizeFault = calcMemSize(stack.peek(), u256(1))
 	case MSTORE:
-		newMemSize = calcMemSize(stack.peek(), u256(32))
+		newMemSize, sizeFault = calcMemSize(stack.peek(), u256(32))
 	case RETURN:
-		newMemSize = calcMemSize(stack.peek(), stack.data[stack.len()-2])
+		newMemSize, sizeFault = calcMemSize(stack.peek(), stack.data[stack.len()-2])
 	case SHA3:
-		newMemSize = calcMemSize(stack.peek(), stack.data[stack.len()-2])
+		newMemSize, sizeFault = calcMemSize(stack.peek(), stack.data[stack.len()-2])
+		if sizeFault {
+			return 0, 0, OutOfGasError
+		}
 
-		words := toWordSize(stack.data[stack.len()-2])
-		gas.Add(gas, words.Mul(words, params.Sha3WordGas))
+		if stack.data[stack.len()-2].BitLen() > 64 {
+			return 0, 0, OutOfGasError
+		}
+		words := toWordSize(stack.data[stack.len()-2].Uint64())
+		if !math.IsMulSafe(words, KeccakWordGas64) {
+			return 0, 0, OutOfGasError
+		}
+		wordsGas := words * KeccakWordGas64
+		if !math.IsAddSafe(gas, wordsGas) {
+			return 0, 0, OutOfGasError
+		}
+		gas += wordsGas
 	case CALLDATACOPY:
-		newMemSize = calcMemSize(stack.peek(), stack.data[stack.len()-3])
+		newMemSize, sizeFault = calcMemSize(stack.peek(), stack.data[stack.len()-3])
+		if sizeFault {
+			return 0, 0, OutOfGasError
+		}
 
-		words := toWordSize(stack.data[stack.len()-3])
-		gas.Add(gas, words.Mul(words, params.CopyGas))
+		words := toWordSize(stack.data[stack.len()-3].Uint64())
+		if !math.IsMulSafe(words, KeccakWordGas64) {
```

Comment:
You are right. They should indeed. Not sure how these ended up here (same goes for your other comments).